# 3유형

소문제 6문제 (문제당 5점)

6회 : 카이제곱검정 / 다중선형회귀    
7회 : 로지스틱회귀 / 다중선형회귀     
8회 : 로지스틱회귀 / 다중선형회귀    


```python


## 통계검정
from scipy.stats import ttest_1samp # 단일표본
from scipy.stats import ttest_ind   # 독립표본
from scipy.stats import ttest_rel   # 대응표본
s,p = ttest_1samp(df1,n)
s,p = ttest_ind(df1,df2)

from scipy.stats import shapiro     # 정규성검정
from scipy.stats import levene      # 등분산검정

from scipy.stats import chi2_contingency # 카이제곱 독립성검정
from scipy.stats import chisquare        # 카이제곱 적합도검정
pd.crosstab(df1,df2) # 범주형 변수 교차테이블


## 그 외 통계검정을 물을 경우
import scipy.stats
dir(scipy.stats) # 관련 패키지 찾아서 데이터 입력

## 분산분석
from scipy.stats import f_oneway # anova f-검정
f_oneway(df1,df2,df3)

import statsmodels.api as sm # 이원 분산 분석
from statsmodels.formula.api import ols # 범주형 변수는 C()로 감싼다
model = ols('Y ~ C(X1) * X2', data=df).fit()
sm.stats.anova_lm(model)

## 다중선형회귀분석
import statsmodels.api as sm
X = sm.add_constant(X) # 상수항 추가
model = sm.OLS(Y, X).fit()
print(model.summary())

## 로지스틱 회귀
# 방법1
from statsmodels.formula.api import logit
import numpy as np
model = logit("종속변수 ~ 독립변수1 + 독립변수2", data=df).fit() # 로지스틱 회귀 분석 모델
print(model.summary())


odds_ratios = np.exp(model.params) # 오즈비
print(odds_ratios)

# 방법2
import statsmodels.api as sm
X = sm.add_constant(X)
model = sm.Logit(y, X).fit() # 로지스틱 회귀 모델 적합
print(model.summary()) # 로지스틱 회귀 모델 적합


# 방법3
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(X,Y)
lr.coef_



odds_ratios = np.exp(model.params) # 오즈비
print(odds_ratios)


-2 * model.llf # 잔차이탈도


```

예시문제 참고 : https://www.datamanim.com/dataset/97_scipy/scipy.html

## 단일표본검정
평균값이 주어지고 데이터가 주어질 때
```python
from scipy.stats import ttest_1samp
귀무가설 (H0): 표본의 평균은 모집단의 평균과 같다.
대립가설 (H1): 표본의 평균은 모집단의 평균과 다르다.
```

## 독립표본검정
독립 표본 검정은 두 표본이 서로 독립적일 때 사용. 두 그룹이 전혀 다른 개체로 구성,아무런 관계가 없는 경우 사용     
남성과 여성의 평균 키 비교
```python
from scipy.stats import ttest_ind
귀무가설 (H0): 두 그룹 간 평균 차이가 없다.
대립가설 (H1): 두 그룹 간 평균 차이가 있다.
```

## 대응표본검정
동일한 대상에 대해 두 번의 측정을 하거나, 같은 그룹에서 두 조건을 비교할 때 사용     
약을 복용하기 전과 후의 환자의 혈압 비교
```python
from scipy.stats import ttest_rel
귀무가설 (H0): 두 표본 간 평균 차이가 없다.
대립가설 (H1): 두 표본 간 평균 차이가 있다.

```

## 정규성검정
Shapiro-Wilk Test
```python
from scipy.stats import shapiro
귀무가설 (H0): 데이터가 정규 분포를 따른다.
대립가설 (H1): 데이터가 정규 분포를 따르지 않는다.

    
그 외
from scipy.stats import kstest    # Kolmogorov-Smirnov kstest(data, 'norm')
from scipy.stats import anderson  # Anderson-Darling   anderson(data, dist='norm')
```

In [ ]:
# sample컬럼 데이터의 정규성을 검정하라
import pandas as pd
df =pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/data/9.csv')
df.head()

sample
0 -0.125684
1 -0.230200
2  0.425670
3  0.707533
4 -0.637835

## 등분산검정

```python
from scipy.stats import levene
귀무가설 (H0): 모든 집단의 분산이 같다 (등분산).
대립가설 (H1): 적어도 하나의 집단의 분산이 다르다 (비등분산).

```

In [ ]:
# 그룹간의 분산에 차이가 있는지 검정하라
import pandas as pd
df =pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/data/10.csv')
df.head()

data group
0  0.535368     A
1  1.060150     B
2 -0.089042     B
3  0.575745     A
4  0.524174     B

## 카이제곱독립성검정

```python
from scipy.stats import chi2_contingency
카이제곱 통계량,p, 자유도, 기대빈도(독립적일 때 각 셀에서 기대되는 빈도)

귀무가설 (H0): 두 변수는 독립적이다. (연관성,상관관계가 없다)
대립가설 (H1): 두 변수는 독립적이지 않다. (연관이 존재한다)
```

In [ ]:
# 성별과 제품선호가 독립인지 확인하라.
import pandas as pd
df =pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/data/11.csv')
df.head()

성별 제품 선호도
0  남성      A
1  남성      A
2  남성      A
3  남성      A
4  남성      A

## 카이제곱적합도검정

```python
from scipy.stats import chisquare

귀무가설 (H0): 관측 데이터는 기대 분포와 같다.
대립가설 (H1): 관측 데이터는 기대 분포와 다르다.
   
# 관측 빈도와 기대 빈도
observed = [30, 10, 20]
expected = [25, 15, 20]

# 적합도 검정
chi2_stat, p_value = chisquare(f_obs=observed, f_exp=expected)
```



## 일원분산분석

```python
from scipy.stats import f_oneway
f_statistic, p_value = f_oneway(a, b, c)
귀무가설(H0): 각 집단의 평균은 모두 같다.
대립가설(H1): 적어도 하나의 집단의 평균이 다르다.    
```

## 이원분산분석

```python


귀무가설 (주효과 1): 첫 번째 독립변수에 따른 평균 차이가 없다.
대립가설 (주효과 1): 첫 번째 독립변수에 따른 평균 차이가 있다.
귀무가설 (주효과 2): 두 번째 독립변수에 따른 평균 차이가 없다.
대립가설 (주효과 2): 두 번째 독립변수에 따른 평균 차이가 있다.
귀무가설 (상호작용 효과): 두 변수 간 상호작용 효과가 없다.
대립가설 (상호작용 효과): 두 변수 간 상호작용 효과가 있다.


import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

# 데이터 생성 (교육 방법과 성별에 따른 성적)
data = {
    '성적': [85, 86, 88, 75, 78, 94, 98, 79, 71, 91, 92, 85, 87, 89, 95, 96, 85, 90],
    '교육방법': ['A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B'],
    '성별': ['남', '여', '남', '여', '남', '여', '남', '여', '남', '남', '여', '남', '여', '남', '여', '남', '여', '남']
}

# DataFrame 생성
df = pd.DataFrame(data)

# 이원분산분석 수행 (교육방법과 성별의 효과 분석)
model = ols('성적 ~ C(교육방법) + C(성별) + C(교육방법):C(성별)', data=df).fit()
anova_table = sm.stats.anova_lm(model, typ=2)

```

In [ ]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

# 데이터 생성 (교육 방법과 성별에 따른 성적)
data = {
    '성적': [85, 86, 88, 75, 78, 94, 98, 79, 71, 91, 92, 85, 87, 89, 95, 96, 85, 90],
    '교육방법': ['A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B'],
    '성별': ['남', '여', '남', '여', '남', '여', '남', '여', '남', '남', '여', '남', '여', '남', '여', '남', '여', '남']
}

# DataFrame 생성
df = pd.DataFrame(data)

# 이원분산분석 수행 (교육방법과 성별의 효과 분석)
model = ols('성적 ~ C(교육방법) + C(성별) + C(교육방법):C(성별)', data=df).fit()
anova_table = sm.stats.anova_lm(model, typ=2)

## 다중선형회귀분석

```python
import statsmodels.api as sm
X = sm.add_constant(X) # 상수항 추가
model = sm.OLS(Y, X).fit()
print(model.summary())

## p값이 0.05이하면 유효한 계수
```

## 로지스틱회귀

```python
# 방법1
from statsmodels.formula.api import logit
import numpy as np
model = logit("종속변수 ~ 독립변수1 + 독립변수2", data=df).fit() # 로지스틱 회귀 분석 모델
print(model.summary())


odds_ratios = np.exp(model.params) # 오즈비
print(odds_ratios)

# 방법2
import statsmodels.api as sm
X = sm.add_constant(X)
model = sm.Logit(y, X).fit() # 로지스틱 회귀 모델 적합
print(model.summary()) # 로지스틱 회귀 모델 적합


# 방법3
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(X,Y)
lr.coef_


## p값이 0.05이하면 유효한 계수
```

In [ ]:
import statsmodels.api as sm
import numpy as np
np.random.seed(42)

# 데이터 생성
data = {
    '공부시간': np.random.randint(1, 10, size=100),  # 1~9 시간
    '수업참석': np.random.randint(0, 2, size=100),   # 0: 안 참석, 1: 참석
    '합격': np.random.randint(0, 2, size=100)         # 0: 불합격, 1: 합격
}

df = pd.DataFrame(data)

# 기출

In [ ]:
# 종속변수 target
import pandas as pd
df =pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/data/12.csv')
df.head()

Target        v1        v2        v3        v4        v5        v6  \
0  21.650072  0.496714 -0.138264  0.647689  1.523030 -0.234153 -0.234137   
1 -19.469855 -0.225776  0.067528 -1.424748 -0.544383  0.110923 -1.150994   
2 -24.475600 -0.115648 -0.301104 -1.478522 -0.719844 -0.460639  1.057122   
3  13.476831 -1.196207  0.812526  1.356240 -0.072010  1.003533  0.361636   
4   3.505322 -0.808494 -0.501757  0.915402  0.328751 -0.529760  0.513267   

         v7        v8        v9  ...       v12       v13       v14       v15  \
0  1.579213  0.767435 -0.469474  ... -0.465730  0.241962 -1.913280 -1.724918   
1  0.375698 -0.600639 -0.291694  ... -0.013497 -1.057711  0.822545 -1.220844   
2  0.343618 -1.763040  0.324084  ...  0.611676  1.031000  0.931280 -0.839218   
3 -0.645120  0.361396  1.538037  ... -2.619745  0.821903  0.087047 -0.299007   
4  0.097078  0.968645 -0.702053  ... -1.463515  0.296120  0.261055  0.005113   

        v16       v17       v18       v19       v20       v21  
0 -0.562288 -1.012831  0.314247 -0.908024 -1.412304  1.465649  
1  0.208864 -1.959670 -1.328186  0.196861  0.738467  0.171368  
2 -0.309212  0.331263  0.975545 -0.479174 -0.185659 -1.106335  
3  0.091761 -1.987569 -0.219672  0.357113  1.477894 -0.518270  
4 -0.234587 -1.415371 -0.420645 -0.342715 -0.802277 -0.161286  

[5 rows x 22 columns]

## 7-1 종속변수를 제외한 데이터에서 v1 컬럼과와 가장 큰 상관계수를 가진 변수를 찾아 그 값을 구하여라

## 7-2 Target 변수를 종속변수로 하여 다중선형회귀모델링을 진행했을 때 v2 컬럼의 회귀 계수는?

## 7-3 회귀 계수들이 가지는 p값들 중 최대 값은?

In [ ]:
# 종속변수 target
import pandas as pd
df =pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/data/13.csv')
df.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca  thal  target  
0    3.0  0.0   6.0       0  
1    2.0  3.0   3.0       1  
2    2.0  2.0   7.0       1  
3    3.0  0.0   3.0       0  
4    1.0  0.0   3.0       0

## 7-4 train 데이터로 target을 종속변수로 로지스틱 회귀를 진행할 때 age 컬럼의 오즈비를 구하여라
(train 데이터는 앞의 210개 행을, test데이터는 나머지 부분을 사용한다)

## 7-5 train으로 로지스틱 회귀 진행했을 경우 잔차 이탈도 (residual deviance)를 계산하라

## 7-6 test 데이터의 독립변수로 target 예측 후 오류율을 구하여라

In [ ]:
# 종속변수 Y
import pandas as pd
df =pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/data/14.csv')
df.head()

X1        X2        X3        X4        X5        X6        X7  \
0  0.506278 -0.231322  0.759771 -0.172927 -0.687524 -0.556815 -1.381369   
1  0.017220  1.530047  0.676838  1.533623 -0.516027  0.949166 -1.180669   
2  1.436089  0.038078 -0.120501  0.914955 -1.504149  0.443893 -0.696385   
3 -0.926337 -0.665242 -0.106853 -1.111627 -0.127867  0.746742 -0.345943   
4 -0.350953 -0.044709 -0.421777  1.359772 -1.619690 -0.645908 -1.717975   

         X8        X9       X10       X11       X12       X13       X14  \
0  1.633690 -0.869129  0.370614  0.615057 -0.416594 -1.175980  0.113494   
1  1.895021 -1.211768 -1.081109 -1.067039 -0.611829  0.846854 -0.812381   
2  0.245013  1.221166 -1.387287 -1.762048 -0.556466  0.130129  0.517210   
3  0.848144  0.374944 -0.951299  1.097359  0.754369 -0.221183 -0.430509   
4 -0.598477  1.026014 -0.134644 -0.988988  0.513048 -0.886113  0.247288   

        X15  Y  
0 -0.539230  1  
1  0.823546  0  
2  1.056730  0  
3  0.756875  1  
4 -0.016664  0

## 8-1 y컬럼을 종속변수로하는 로지스틱회귀 모형을 만들 떄 유의미하지 않은 변수의 개수는?

## 8-2 유의한 변수들 만으로 다시 모델링을 해서 독립변수(상수항제외)들의 회귀계수 합계를 소숫점 이하 3째 자리까지 구하여라


## 8-3 위 모델에서 사용된 독립변수 중 'X' 뒤의 숫자가 가장 낮은 컬럼이 5 단위가 증가할시 오즈비는 몇 배로 변화하는가?

## 8-4 PIQ를 예측하는 다중선형회귀 모형을 만들고 Brain의 회귀 계수를 구하여라

## 8-5 모델의 r2를 구하여라

## 8-6 위에서 계산한 모델을 이용하여 뇌크기 90 키 70 몸무게 150일 때 지능을 예측하라

# 9-1
- 아래 데이터는 각 학생(id)의 디자인 시험 점수(design)를 나타낸다.
- f1~f5는 5가지의 수행평가 점수를 나타낸다.
- id값이 294번까지는 train 나머지는 Test 데이터로 사용한다


id컬럼을 제외한 모든 변수를 활용해 train 데이터에 대해 시험점수를 예측하는 다중회귀분석을 시행한 후 (절편포함,패널티 x) 유의하지 않은 설명변수 개수를 구하라

In [ ]:
import pandas as pd
df =pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/refs/heads/master/krdatacertificate/10/s1.csv')
df.head()

id        f1        f2        f3        f4        f5      design
0   1  3.745401  8.124211  1.117850  3.420941  2.384399  100.000000
1   2  9.507143  8.106563  6.971836  8.178032  4.902441  100.000000
2   3  7.319939  1.820855  2.602324  3.793258  3.563888   79.022594
3   4  5.986585  5.449783  9.649633  5.096981  4.289922   97.220618
4   5  1.560186  1.518029  2.337965  1.104585  1.530099   59.873376

# 9-2
- 9-1에서 구한 유의한 변수만 사용하여 다시 다중회귀분석을 진행하고(절편 포함) train 데이터를 사용하여 해당 모델의 예측값과 실제값의 피어슨 상관계수를 구하여라.

# 9-3
- 9-2에서 생성된 모델을 사용해 test 데이터에 대한 rmse를 구하여라.

# 9-4
아래 데이터는 핸드폰 이용정보를 담고 있고 이탈(churn)유무를 판단하기 위한 데이터셋이다.

- 로지스틱 회귀를 진행(절편 추가)하고 변수 avg_call_duration에 대한 p-value를 구하시오.

In [ ]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/refs/heads/master/krdatacertificate/10/s2.csv')
df.head()

daily_usage  monthly_charge  support_contacts  data_usage  \
0           59        8.402839                 8    7.933944   
1           14        8.596278                 2    6.074770   
2           21        9.924368                 9    5.721576   
3           62        3.766421                10   23.670044   
4           36        8.036970                 2   29.704139   

   avg_call_duration  age  churn  
0           4.685791   42      0  
1           3.164515   53      1  
2          10.495208   48      1  
3           8.915550   64      0  
4           8.381676   24      0

# 9-5
다른 변수의 영향은 무시하고 data_usage 변수가 1에서 3이 됐을 때의 오즈비를 구하라

# 9-6
생성된 모델에 데이터를 넣어 고객이 이탈 할 확률값을 출력한 다음 그 확률값이 0.4보다 큰 고객의 수를 계산하라.